### ISSUE TITLE
REGEXP_REPLACE with lambda triggers parsing errors in Trino Dialect

### ISSUE THREAD

**Issue by rileymcdowell** [sqlfluff/sqlfluff#5582]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

I wrote some Trino SQL with a `REGEXP_REPLACE(x, y, z -> z[0])` in the query and linted it with SQLFluff. This generated a parsing error.

Upon further investigation, I found the other forms of REGEXP_REPLACE were parseable, but the one involving the lambda definition token `->` isnot.

### Expected Behaviour

I expect `REGEXP_REPLACE(x, y, z -> z[0])` to be parsable as valid syntax by SQLFluff.

### Observed Behaviour

Here's an example SQL file. I stole this example from the Trino regexp documentation [here](https://trino.io/docs/current/functions/regexp.html).

```sql
-- Capitalization method implemented via regex+lambda.
SELECT REGEXP_REPLACE('new york', '(\w)(\w*)', x -> UPPER(x[1]) || LOWER(x[2]));
```
Here's my SQLFluff command:

```bash
sqlfluff lint --dialect trino ./query.sql
```

And here's the output:

```txt
== [query.sql] FAIL
L:   1 | P:  50 |  PRS | Line 1, Position 50: Found unparsable section: '->
                       | UPPER(x[1]) || LOWER(x[2])'
WARNING: Parsing errors found and dialect is set to 'trino'. Have you configured your dialect correctly?
All Finished 📜 🎉!
```

### How to reproduce

Create a one-line SQL file called `query.sql` with the following contents:

```sql
SELECT REGEXP_REPLACE('new york', '(\w)(\w*)', x -> UPPER(x[1]) || LOWER(x[2]))
```

Then run the following command with the latest SQLFluff release:
```bash
sqlfluff lint --dialect trino ./query.sql
```

### Dialect

Trino

### Version

```bash
sqlfluff --version
sqlfluff, version 2.3.5
```

```bash
python --version
Python 3.9.16
```

### Configuration

No configuration file was used. This can be triggered with the default configuration.

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
